--- a/dwl.c
+++ b/dwl.c
@@ -319,6 +319,7 @@
 static void rendermon(struct wl_listener *listener, void *data);
 static void requestdecorationmode(struct wl_listener *listener, void *data);
 static void requeststartdrag(struct wl_listener *listener, void *data);
+static void runautostart(void);
 static void requestmonstate(struct wl_listener *listener, void *data);
 static void resize(Client *c, struct wlr_box geo, int interact);
 static void run(char *startup_cmd);
@@ -359,6 +360,8 @@
 /* variables */
 static const char broken[] = "broken";
 static pid_t child_pid = -1;
+static pid_t *autostart_pids;
+static size_t autostart_len;
 static int locked;
 static void *exclusive_focus;
 static struct wl_display *dpy;
@@ -685,6 +688,11 @@
 		kill(-child_pid, SIGTERM);
 		waitpid(child_pid, NULL, 0);
 	}
+	for (size_t i = 0; i < autostart_len; i++) {
+		if (autostart_pids[i] > 0)
+			kill(autostart_pids[i], SIGTERM);
+	}
+	free(autostart_pids);
 	wlr_xcursor_manager_destroy(cursor_mgr);
 
 	destroykeyboardgroup(&kb_group->destroy, NULL);
@@ -2224,6 +2232,9 @@
 	if (!wlr_backend_start(backend))
 		die("startup: backend_start");
 
+	/* Run autostart programs now that WAYLAND_DISPLAY is set */
+	runautostart();
+
 	/* Now that the socket exists and the backend is started, run the startup command */
 	if (startup_cmd) {
 		int piperw[2];
@@ -2656,6 +2667,34 @@
 	}
 }
 
+void
+runautostart(void)
+{
+	const char *const *p;
+
+	/* Count number of commands */
+	autostart_len = 0;
+	for (p = autostart; *p; autostart_len++) {
+		for (; *p; p++)
+			;
+		p++;
+	}
+
+	autostart_pids = calloc(autostart_len, sizeof(*autostart_pids));
+
+	size_t i = 0;
+	for (p = autostart; *p; i++) {
+		if ((autostart_pids[i] = fork()) == 0) {
+			setsid();
+			execvp(*p, (char *const *)p);
+			die("dwl: autostart execvp %s failed:", *p);
+		}
+		for (; *p; p++)
+			;
+		p++;
+	}
+}
+
 void
 startdrag(struct wl_listener *listener, void *data)
 {
